File: app/schemas/city.py
from pydantic import BaseModel, Field
from typing import Optional, List, Dict
from uuid import UUID
from datetime import datetime


class NeighborhoodBase(BaseModel):
    """Base schema for neighborhood."""
    name: str
    lat: float
    lon: float
    vibes: List[str] = []
    best_for: List[str] = []


class NeighborhoodScore(NeighborhoodBase):
    """Neighborhood with persona score."""
    score: float = Field(ge=0, le=1)
    reasoning: str


class BoundingBox(BaseModel):
    """Geographic bounding box."""
    min_lon: float
    min_lat: float
    max_lon: float
    max_lat: float


class CityCenter(BaseModel):
    """City center coordinates."""
    lat: float
    lon: float


class CityBase(BaseModel):
    """Base city schema."""
    name: str
    country: str
    currency: str
    bbox: BoundingBox
    center: CityCenter
    neighborhoods: List[NeighborhoodBase] = []


class CityCreate(BaseModel):
    """Schema for creating a new city."""
    name: str
    country: str
    currency: str
    min_lon: float
    min_lat: float
    max_lon: float
    max_lat: float
    neighborhoods: Optional[List[NeighborhoodBase]] = None


class CityResponse(CityBase):
    """City response schema."""
    id: str  # city key (e.g., "paris")
    landmark_count: Optional[int] = None
    poi_count: Optional[int] = None


class CityListResponse(BaseModel):
    """Response for list of cities."""
    cities: List[CityResponse]
    total: int


class NeighborhoodRecommendRequest(BaseModel):
    """Request for neighborhood recommendations."""
    city: str
    group_type: str  # family, couple, solo, friends, honeymoon, seniors, business
    vibes: List[str]  # cultural, romantic, foodie, etc.


class NeighborhoodRecommendResponse(BaseModel):
    """Response for neighborhood recommendations."""
    city: str
    group_type: str
    recommendations: List[NeighborhoodScore]

File: flutter_app/lib/models/city.dart
class City {
  final String id;
  final String name;
  final String country;
  final String currency;
  final BoundingBox? bbox;
  final Coordinates? center;
  final int? landmarkCount;
  final String? imageUrl;

  City({
    required this.id,
    required this.name,
    required this.country,
    this.currency = 'EUR',
    this.bbox,
    this.center,
    this.landmarkCount,
    this.imageUrl,
  });

  factory City.fromJson(Map<String, dynamic> json) {
    return City(
      id: json['id'] ?? '',
      name: json['name'] ?? '',
      country: json['country'] ?? '',
      currency: json['currency'] ?? 'EUR',
      bbox: json['bbox'] != null ? BoundingBox.fromJson(json['bbox']) : null,
      center: json['center'] != null ? Coordinates.fromJson(json['center']) : null,
      landmarkCount: json['landmark_count'],
      imageUrl: json['image_url'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'country': country,
      'currency': currency,
      'bbox': bbox?.toJson(),
      'center': center?.toJson(),
      'landmark_count': landmarkCount,
      'image_url': imageUrl,
    };
  }
}

class BoundingBox {
  final double minLon;
  final double maxLon;
  final double minLat;
  final double maxLat;

  BoundingBox({
    required this.minLon,
    required this.maxLon,
    required this.minLat,
    required this.maxLat,
  });

  factory BoundingBox.fromJson(Map<String, dynamic> json) {
    return BoundingBox(
      minLon: (json['min_lon'] ?? 0).toDouble(),
      maxLon: (json['max_lon'] ?? 0).toDouble(),
      minLat: (json['min_lat'] ?? 0).toDouble(),
      maxLat: (json['max_lat'] ?? 0).toDouble(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'min_lon': minLon,
      'max_lon': maxLon,
      'min_lat': minLat,
      'max_lat': maxLat,
    };
  }
}

class Coordinates {
  final double lat;
  final double lon;

  Coordinates({required this.lat, required this.lon});

  factory Coordinates.fromJson(Map<String, dynamic> json) {
    return Coordinates(
      lat: (json['lat'] ?? json['latitude'] ?? 0).toDouble(),
      lon: (json['lon'] ?? json['longitude'] ?? 0).toDouble(),
    );
  }

  Map<String, dynamic> toJson() {
    return {'lat': lat, 'lon': lon};
  }
}
